File: 프로그래머스/lv2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

char words[5] = {'A', 'E', 'I', 'O', 'U'};
vector<string> dic;

void comb(string& res, int lim, int next) {
    res.push_back(words[next]);
    if (lim == res.size()) {
        dic.push_back(res);
        res.pop_back();
        return;
    }
    for (int i = 0; i < 5; i++)
        comb(res, lim, i);
    res.pop_back();
}

#include <iostream>
int solution(string word) {
    string tmp;
    for (int i = 1; i <= 5; i++) {
        for (int j = 0; j < 5; j++)
            comb(tmp, i, j);
    }
    sort(dic.begin(), dic.end());
    for (int i = 0; i < dic.size(); i++)
        if (dic[i] == word) return i+1;
    return 0;
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
class Solution {
    
    private int getIndex(char c) {
        switch (c) {
            case 'A': return 0;
            case 'E': return 1;
            case 'I': return 2;
            case 'O': return 3;
            case 'U': return 4;
        }
        return -1;
    }
    
    public int solution(String word) {
        int[] cnts = new int[]{781, 156, 31, 6, 1};
        int res = word.length();
        for (int i = 0; i < word.length(); i++)
            res += getIndex(word.charAt(i)) * cnts[i];
        return res;
    }
}